lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+                           getIncrementalIndex(),
+                           CurrentSuccs });
+      }
+    }
+  }
+}
+
+/// Helper function that returns only the `forward` edges, according to
+/// the spanning tree. Forward edges are defined as edges connecting the current
+/// `Source` node to nodes that are successor of it on the spanning tree.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getForwardEdges(BlockNode *Source) {
+  EdgesVectorType ForwardEdges;
+
+  // We save the size of the successors when the graph is not treated as
+  // undirected (we use this info to understand when the concatenated iterator
+  // has surpassed the end of the non-concat one, so we can deduce if we are
+  // seeing the inverted version of the edge wrt. the directed version of the
+  // graph)
+  size_t EndOfSuccs = llvm::size(llvm::children_edges<BlockNode *>(Source));
+  size_t Index = 0;
+
+  for (auto &[Neighbor, Label] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Source,
+                                                     Neighbor,
+                                                     Label->id() });
+    if (Label->type() == EdgeType::TreeEdge and not isBackedge(Edge)) {
+
+      // Compute the `IsInverted` information checking if our iterator superated
+      // the end of the successors (and ended up in the predecessor half
+      // computed by the concatenation)
+      bool IsInverted = Index >= EndOfSuccs;
+      ForwardEdges.push_back({ Edge, Label, IsInverted });
+    }
+
+    Index++;
+  }
+  return ForwardEdges;
+}
+
+/// Helper function that returns only the `backward` edges, according to the
+/// spanning tree. Backward edges are defined as edges connecting the current
+/// `Source` node to nodes that are not successors of it on the spanning tree,
+/// and therefore are backedges.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getBackwardEdges(BlockNode *Source) {
+  EdgesVectorType BackwardEdges;
+
+  // We save the size of the successors when the graph is not treated as
+  // undirected (we use this info to understand when the concatenated iterator
+  // has surpassed the end of the non-concat one, so we can deduce if we are
+  // seeing the inverted version of the edge wrt. the directed version of the
+  // graph)
+  size_t EndOfSuccs = llvm::size(llvm::children_edges<BlockNode *>(Source));
+  size_t Index = 0;
+
+  for (auto &[Neighbor, Label] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Source,
+                                                     Neighbor,
+                                                     Label->id() });
+    if (Label->type() == EdgeType::BackEdge and isBackedge(Edge)) {
```

Comment:
Can it ever happen that the LHS of `and` is `true` and the RHS `false`??

Or should `isBackedge(Edge)` be an assertion inside the `if`?

Reply:
Actually this happens for all the other uses of `isBackedge`. It seems to me that the use of `isBackedge` (or `not isBackedge`) in other occurrences is redundant and should be turned into an assertion.

If not, then there's something I'm missing.

Reply:
Perhaps the things I'm missing have to do with self-loops?

Reply:
The terminology is not clear and may lead to confusion, but the LHS and the RHS refers to two very different things. The LHS refers to the fact that the edge on the undirected graph has been marked as a backedge on the spanning tree. The RHS (the invocation to the `isBackedge` helper function) instead, just checks the `DFSNum`s, and tells that a specific edge is a backedge if the `DFSNum` of the source node is greater than the `DFSNum` of the target.

Both the criteria are needed, even if they may seem redundant, and this is due to the fact that on directed graph we can have tree edges, back edges, forward edges and cross edges, while on undirected graphs we can simply have tree edges and back edges. Due to this, the detection of forward and backedges needs both these criteria.

However, with the latest changes here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L321, the above check is now:
```
if (Label->type() == SpanningTreeEdgeKind::BackEdge and isBackedge(Edge)) {
```
which highlights more clearly the disction between the criteria

Reply:
Oh ok it's much clearer now.